internal/controller/adapters/sveltos/serviceset_controller.go

```diff
@@ -0,0 +1,1135 @@
+	errBuildPolicyRefsFailed        = errors.New("failed to build policy refs")
+)
+
+type ProfileConfig struct {
+	// KSM specific configuration
+	// Priority is the priority of the Profile.
+	Priority *int32 `json:"priority,omitempty"`
+	// DriftIgnore is a list of [github.com/projectsveltos/libsveltos/api/v1beta1.PatchSelector] to ignore
+	// when checking for drift.
+	DriftIgnore []libsveltosv1beta1.PatchSelector `json:"driftIgnore,omitempty"`
+
+	SyncMode             string                                       `json:"syncMode,omitempty"`
+	TemplateResourceRefs []addoncontrollerv1beta1.TemplateResourceRef `json:"templateResourceRefs,omitempty"`
+	DriftExclusions      []addoncontrollerv1beta1.DriftExclusion      `json:"driftExclusions,omitempty"`
+	Patches              []libsveltosv1beta1.Patch                    `json:"patches,omitempty"`
+	ContinueOnError      bool                                         `json:"continueOnError,omitempty"`
+	Reloader             bool                                         `json:"reloader,omitempty"`
+	StopOnConflict       bool                                         `json:"stopOnConflict,omitempty"`
+}
+
+// ServiceSetReconciler reconciles a ServiceSet object and produces
+// [github.com/projectsveltos/addon-controller/api/v1beta1.Profile] objects.
+type ServiceSetReconciler struct {
+	client.Client
+
+	timeFunc func() time.Time
+
+	// AdapterName is the name of the workload running the controller
+	// effectively this name is used to identify adapter in the
+	// [github.com/k0rdent/kcm/api/v1beta1.StateManagementProvider] spec.
+	AdapterName      string
+	AdapterNamespace string
+
+	requeueInterval time.Duration
+}
+
+func (r *ServiceSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	start := time.Now()
+	l := ctrl.LoggerFrom(ctx)
+	l.Info("Reconciling ServiceSet")
+
+	serviceSet := new(kcmv1.ServiceSet)
+	err = r.Get(ctx, req.NamespacedName, serviceSet)
+	if apierrors.IsNotFound(err) {
+		l.Info("ServiceSet not found, skipping")
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if !serviceSet.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, serviceSet)
+	}
+
+	if !controllerutil.ContainsFinalizer(serviceSet, kcmv1.ServiceSetFinalizer) {
+		controllerutil.AddFinalizer(serviceSet, kcmv1.ServiceSetFinalizer)
+		return ctrl.Result{}, r.Update(ctx, serviceSet)
+	}
+
+	smp := new(kcmv1.StateManagementProvider)
+	if err = r.Get(ctx, client.ObjectKey{Name: serviceSet.Spec.Provider.Name}, smp); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	continueReconciliation, err := labelsMatchSelector(serviceSet.Labels, smp.Spec.Selector)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to check ServiceSet labels: %w", err)
+	}
+	if !continueReconciliation {
+		l.V(1).Info("ServiceSet labels do not match provider selector, skipping")
```

Comment:
For my own understanding, could you kindly share a scenario where this could happen?